### DISCUSSION TITLE
How to build into a single ESM module for offline use?

### DISCUSSION THREAD

**Discussion by datvm** [material-components/material-web#4892] [Q&A] [answered]
I am not familiar with Node and its technologies and have been trying to use Rollup or Webpack to pack the whole thing into a single JS file without any success. Could someone please give me some guide on this?

I want a (preferrably a single file) build so I can refer to it in an HTML file. No server side, just a WebView app so I need everything offline. Similar to how right now you can refer to https://unpkg.com/@material/web@1.0.0-pre.17/all.js?module or https://unpkg.com/@material/web@1.0.0-pre.17/common.js?module

**Comment by Aworldc** [comment, marked as answer]
try using [esbuild](https://esbuild.github.io/) with the following options:
```bash
$ esbuild --bundle --minify --format=esm --outfile=bundle.js "path/to/all.js"
```

**Reply by datvm** [reply]
Thanks, turned out I still need to use their command to build the SCSS files. I couldn't run it in Windows without WSL. Your esbuild command works!

**Reply by datvm** [reply]
I just tried it in an HTML page, it works. Now I just need to find a way to download all the Fonts for Google Sans and Symbol for offline use.

**Reply by Aworldc** [reply]
ah yes. fonts are always a pain point.

**Comment by asyncliz** [comment]
We also have a [quick start](https://github.com/material-components/material-web/blob/main/docs/quick-start.md) that includes a building section!

If you saw that but had problems, please let us know! We'll want to fix the guide if so.

**Reply by datvm** [reply]
I tried that one at first (even running `npm run build`) but unfortunately that script doesn't work on Windows (missing Linux commands). Strangely I tried running it with an equivalent Powershell command instead but no CSS file was generated. I had to run it with WSL to get the SCSS and TypeScript compiled and then use the above answer `esbuild` to pack all the JS files into one bundle.

**Reply by datvm** [reply]
@asyncLiz I created this [GitHub Action](https://github.com/datvm/material-web/blob/main/.github/workflows/build-es-bundle.yml) to build and bundle the component into a single ESM package. It's really simple on Linux environment really. The minified file is 441KB, very good 👍 It's also surprisingly much faster than running on my Windows WSL, it took about 2-3 minutes for compiling SASS vs a few seconds on Github Actions.
